expression/builtin_json.go

```diff
@@ -688,8 +690,67 @@ type jsonValidFunctionClass struct {
 	baseFunctionClass
 }
 
-func (c *jsonValidFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
-	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "JSON_VALID")
+func (c *jsonValidFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (sig builtinFunc, err error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var bf baseBuiltinFunc
+	switch args[0].GetType().EvalType() {
+	case types.ETJson:
+		bf = newBaseBuiltinFuncWithTp(ctx, args, types.ETInt, types.ETJson)
+		sig = &builtinJSONValidJSONSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_JsonValidJsonSig)
+	default:
+		bf = newBaseBuiltinFuncWithTp(ctx, args, types.ETInt, args[0].GetType().EvalType())
+		sig = &builtinJSONValidStringSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_JsonValidStringSig)
+	}
+	return sig, nil
+}
+
+type builtinJSONValidJSONSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinJSONValidJSONSig) Clone() builtinFunc {
+	newSig := &builtinJSONValidJSONSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinJSONValidJSONSig) evalInt(row chunk.Row) (res int64, isNull bool, err error) {
+	_, isNull, err = b.args[0].EvalJSON(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, nil
+	}
+	return 1, false, nil
+}
+
+type builtinJSONValidStringSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinJSONValidStringSig) Clone() builtinFunc {
+	newSig := &builtinJSONValidStringSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinJSONValidStringSig) evalInt(row chunk.Row) (res int64, isNull bool, err error) {
+	if b.args[0].GetType().EvalType() != types.ETString {
```

Comment:
Why don't we cast string for arg[0] in func (c *jsonValidFunctionClass) getFunction().
You may refer the comment and implementation of `newBaseBuiltinFuncWithTp`.

Reply:
We should not cast arg[0] in func (c *jsonValidFunctionClass) getFunction() to string, because if we did this, it will cause `select json_valid(1);` not compatible with mysql.
![image](https://user-images.githubusercontent.com/41641871/49053631-dcfec600-f22b-11e8-9246-1d5859ef116d.png)

Reply:
``` go
func (c *jsonUnquoteFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
	if err := c.verifyArgs(args); err != nil {
		return nil, errors.Trace(err)
	}
	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETString, types.ETJson)
	args[0].GetType().Flag &= ^mysql.ParseToJSONFlag // Should we also set this flag when wrap a cast?
	sig := &builtinJSONUnquoteSig{bf}
	sig.setPbCode(tipb.ScalarFuncSig_JsonUnquoteSig)
	return sig, nil
}
```

Reply:
Sorry, i have add this line `args[0].GetType().Flag &= ^mysql.ParseToJSONFlag // Should we also set this flag when wrap a cast?`. But the result has no change. And why must we cast string for  arg[0] in func (c *jsonValidFunctionClass) getFunction()?